File: bindings/python/tests/test_api.py
#!/usr/bin/env python3
"""
Simple test script for API version and library initialization.

Run with: python3 tests/test_api.py
"""

import sys
import os

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import cc


def test_api_version():
    """Test API version functions."""
    print("TEST: cc.api_version()...", end=" ")
    version = cc.api_version()
    assert version == "0.1.0", f"Expected 0.1.0, got {version}"
    print("PASSED")


def test_api_version_compatible():
    """Test version compatibility checking."""
    print("TEST: cc.api_version_compatible()...", end=" ")
    assert cc.api_version_compatible(0, 1), "0.1 should be compatible"
    assert cc.api_version_compatible(0, 0), "0.0 should be compatible"
    assert not cc.api_version_compatible(1, 0), "1.0 should NOT be compatible"
    assert not cc.api_version_compatible(0, 99), "0.99 should NOT be compatible"
    print("PASSED")


def test_init_shutdown():
    """Test library initialization and shutdown."""
    print("TEST: cc.init() / cc.shutdown()...", end=" ")
    cc.init()
    cc.shutdown()
    print("PASSED")


def test_guest_protocol_version():
    """Test guest protocol version."""
    print("TEST: cc.guest_protocol_version()...", end=" ")
    cc.init()
    try:
        version = cc.guest_protocol_version()
        assert version == 1, f"Expected 1, got {version}"
    finally:
        cc.shutdown()
    print("PASSED")


def main():
    print("=== API Tests ===\n")

    test_api_version()
    test_api_version_compatible()
    test_init_shutdown()
    test_guest_protocol_version()

    print("\n=== All API tests passed! ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: bindings/rust/tests/test_api.rs
//! API version and library initialization tests.
//!
//! These tests do not require hypervisor access.

use cc;

#[test]
fn test_api_version() {
    let version = cc::api_version();
    assert_eq!(version, "0.1.0", "Expected API version 0.1.0, got {}", version);
}

#[test]
fn test_api_version_compatible() {
    // Compatible versions
    assert!(
        cc::api_version_compatible(0, 1),
        "0.1 should be compatible"
    );
    assert!(
        cc::api_version_compatible(0, 0),
        "0.0 should be compatible"
    );

    // Incompatible versions
    assert!(
        !cc::api_version_compatible(1, 0),
        "1.0 should NOT be compatible"
    );
    assert!(
        !cc::api_version_compatible(0, 99),
        "0.99 should NOT be compatible"
    );
}

#[test]
fn test_init_shutdown() {
    cc::init().expect("init should succeed");
    cc::shutdown();
}

#[test]
fn test_guest_protocol_version() {
    cc::init().expect("init should succeed");
    let version = cc::guest_protocol_version();
    assert_eq!(version, 1, "Expected guest protocol version 1, got {}", version);
    cc::shutdown();
}

#[test]
fn test_supports_hypervisor() {
    cc::init().expect("init should succeed");
    // This should not panic regardless of whether hypervisor is available
    let result = cc::supports_hypervisor();
    assert!(result.is_ok(), "supports_hypervisor should return Ok");
    cc::shutdown();
}

#[test]
fn test_query_capabilities() {
    cc::init().expect("init should succeed");
    let caps = cc::query_capabilities().expect("query_capabilities should succeed");

    // Architecture should be non-empty
    assert!(
        !caps.architecture.is_empty(),
        "architecture should not be empty"
    );

    // Should be one of the known architectures
    assert!(
        caps.architecture == "x86_64"
            || caps.architecture == "amd64"
            || caps.architecture == "arm64"
            || caps.architecture == "aarch64"
            || caps.architecture == "riscv64",
        "unexpected architecture: {}",
        caps.architecture
    );

    println!(
        "Capabilities: hypervisor={}, arch={}, max_memory={}MB, max_cpus={}",
        caps.hypervisor_available, caps.architecture, caps.max_memory_mb, caps.max_cpus
    );

    cc::shutdown();
}
